### PR TITLE
bcrypt for Node.js with Node addon api

### LINKED ISSUE #513: N-API Support for node.bcrypt.js
The recently announced Node 8 has a new experimental feature called [N-API]( https://nodejs.org/dist/latest-v8.x/docs/api/n-api.html) which is aimed at reducing maintenance cost for node native addons. Checkout this [blog]( https://medium.com/@nodejs/n-api-next-generation-node-js-apis-for-native-modules-169af5235b06) for more details on its benefits. 

`node.bcrypt.js` is one of the top 30 native modules by dependency count, and in order to help the Node.js community make this important transition to N-API, we are hoping you will be able to work with us in order to port `node.bcrypt.js` to support N-API.  Your support and feedback is important in making this effort a success.
 
I am part of the  N-API working group and and would like to talk to you about how you can can get started with N-API and what help we might be able to provide.  I'm available to talk on the phone individually if you'd like.  Alternatively, if you prefer, feel free to jump in our [WG meeting](https://plus.google.com/u/0/events/c0eevtrlajniu7h8cjrdk0f56c8?authkey=COH04YCalJS8Ug) which happens every Thursday at *1.30 Eastern / 10.30 Pacific US time*, to discuss any issues. 
 
Here’s a video of N-API in action 
https://www.youtube.com/watch?v=nmXhJ88nZsk

### PR DESCRIPTION
**Pull request by NickNaso**
Hi,
I'm opening the PR to start iterate over the porting of bcrypt to N-API

Closes #513

### CODE REVIEW

File: src/bcrypt_node.cc [hunk 1]
```diff
@@ -63,86 +60,71 @@ bool ValidateSalt(const char* salt) {
...
 }
 
-NAN_METHOD(GenerateSaltSync) {
-    Nan::HandleScope scope;
-
+Napi::Value GenerateSaltSync (const Napi::CallbackInfo& info) {
     if (info.Length() < 2) {
-        Nan::ThrowTypeError("2 arguments expected");
-        return;
+        throw Napi::TypeError::New(info.Env(), "3 arguments expected");
```
Comment (agathver) [inline src/bcrypt_node.cc]: should be 2 args expected ?

File: binding.gyp [hunk 2]
```diff
@@ -8,8 +8,15 @@
...
       'include_dirs' : [
-          "<!(node -e \"require('nan')\")"
+          "<!@(node -p \"require('node-addon-api').include\")"
       ],
+      'msvs_settings': {
+        'VCCLCompilerTool': { 'ExceptionHandling': 1 },
+      },
+      "xcode_settings": {
+        "CLANG_CXX_LIBRARY": "libc++",
+        'GCC_ENABLE_CPP_EXCEPTIONS': 'YES'
```
Comment (agathver) [inline binding.gyp]: We need same settings in linux too.

**Comment by agathver** [comment]
@NickNaso The windows targets fail with nasty errors. Can you have a look ?

File: src/bcrypt_node.cc (hunk 1 above)
Comment (NickNaso) [reply src/bcrypt_node.cc]: Ok I think that I’m the code there are some errors. I will change as soon as possible now I’m leaving from Vancouver to Italy and I cannot be very active

**Comment by NickNaso** [comment]
@agathver  I have to do others little modifications. Can I add me to the contributors?

File: package.json [hunk 3]
```diff
@@ -29,7 +29,7 @@
     "install": "node-pre-gyp install --fallback-to-build"
   },
   "dependencies": {
-    "nan": "2.6.2",
+    "node-addon-api": "git+https://github.com/nodejs/node-addon-api.git",
```
Comment (defunctzombie) [inline package.json]: Do not merge this. This is a dependency on a floating master branch and will lead to breaking changes. Please pick a specific version to depend on.

Comment (NickNaso) [reply package.json]: Yes you're right I want only understand where was the real problem if on the code or on ```node-addon-api`` version

**Comment by defunctzombie** [comment]
Is `nan` discontinued? Looking at the node-addon-api project, it seems to serve the exact same purpose as `nan`. What is the benefit of switching?

**Comment by agathver** [comment]
@defunctzombie See this: 
- https://medium.com/the-node-js-collection/n-api-next-generation-node-js-apis-for-native-modules-169af5235b06
- https://nodejs.org/api/addons.html

In short, this will bring "compile once, run in any node" versions of native addons, which will save dozens of "bcrypt does not work" issues on the release of a new node version.

**Comment by NickNaso** [comment]
Hi @agathver @defunctzombie 
```node-addon-api``` and ```nan``` are very different. **NAN** could be considered a thin layer build over the V8 API, so it's strong bounded with it. The main purpose of **NAN** is to guarantee the API compatibility across different version of Node.js.
```node-addon-api``` is a C++ wapper of the new **N-API**  [https://nodejs.org/api/n-api.html](https://nodejs.org/api/n-api.html) it's completely JavaScript engine agnostic and it guarentees the API and ABI compatibility between different version of Node.js. SO if you switch to a different version of Node.js you must not reinstall and maybe recompile the addon. 
You could consider the idea to have two package one with N-API and other with NAN as reported here:
[https://nodejs.org/en/docs/guides/publishing-napi-modules/](https://nodejs.org/en/docs/guides/publishing-napi-modules/)

**Comment by defunctzombie** [comment]
`It is intended to insulate Addons from changes in the underlying JavaScript engine and allow modules compiled for one version to run on later versions of Node.js without recompilation.`

This is a cute goal that I doubt will be achieved; maybe if they never introduce backwards breaking abi changes but with c++ that can be quite hard. In any case, it seems like this is the forward direction nodejs itself is promoting instead of `nan`. Thanks for doing the porting.

**Comment by NickNaso** [comment]
Hi @agathver @defunctzombie ,
today we have been released the new version of **node-addon-api**. I substituted it on ```package.json``` now all seems work fine. 
What do you think if with this starting point we work together to publish a release of bcrypt that use N-API?
I mean to follow this guide https://nodejs.org/en/docs/guides/publishing-napi-modules/ obviously before that we need to do all work that you feel necessary to publish the addon.

**Comment by ncb000gt** [comment]
@NickNaso please feel free to add yourself to the list of contributors.

File: src/bcrypt_node.cc [hunk 4]
```diff
@@ -63,86 +60,72 @@ bool ValidateSalt(const char* salt) {
...
 
-NAN_METHOD(GenerateSalt) {
-    Nan::HandleScope scope;
-
+Napi::Value GenerateSalt(const Napi::CallbackInfo& info) {
     if (info.Length() < 3) {
-        Nan::ThrowTypeError("3 arguments expected");
-        return;
+        throw Napi::TypeError::New(info.Env(), "3 arguments expected");
+        return info.Env().Undefined();
```
Comment (ncb000gt) [inline src/bcrypt_node.cc]: Why are these returns needed if you're throwing? I see them throughout, but thought they were unnecessary, no?

Comment (NickNaso) [reply src/bcrypt_node.cc]: You're right it's not necessary if exceptions system is enable. I provided to remove from the code all unuseful code.

File: package.json [hunk 5]
```diff
@@ -29,8 +29,9 @@
     "install": "node-pre-gyp install --fallback-to-build"
   },
   "dependencies": {
-    "nan": "2.6.2",
-    "node-pre-gyp": "0.6.36"
+    "bindings": "^1.3.0",
```
Comment (agathver) [inline package.json]: Please pin these dependencies.

Comment (agathver) [reply package.json]: No. Just remove the caret. "^1.3.0" becomes "1.3.0" and so on.

Comment (NickNaso) [reply package.json]: Ok just done. Can you check?

**Comment by NickNaso** [comment]
Hi guys,
can you continue to review the code so we can publish the N-API version of bcrypt

**Comment by agathver** [comment]
@defunctzombie @ncb000gt Thoughts?

**Comment by NickNaso** [comment]
Hi guys,
what do you think if we tag the version as **1.0.3-napi** and publish it throw this command:
```
npm publish --tag n-api
```
More explanation here:
[To publish N-API version of a package alongside a non-N-API version](https://nodejs.org/en/docs/guides/publishing-napi-modules/)

**Pull request merged**